File: leetcode/problems/problem-specific/290.word-pattern/290.word-pattern.go
package main

import "strings"

func wordPattern(pattern string, s string) bool {
	sSplit := strings.Split(s, " ")
	if len(pattern) != len(sSplit) {
		return false
	}

	patternMappings := map[byte]string{}
	sMappings := map[string]byte{}
	for i := range pattern {
		_, ok1 := patternMappings[pattern[i]]
		_, ok2 := sMappings[sSplit[i]]

		if !ok1 && !ok2 {
			patternMappings[pattern[i]] = sSplit[i]
			sMappings[sSplit[i]] = pattern[i]
		} else if patternMappings[pattern[i]] != sSplit[i] || sMappings[sSplit[i]] != pattern[i] {
			return false
		}
	}

	return true
}

File: leetcode/problems/problem-specific/290.word-pattern/290.word-pattern.py
#
# @lc app=leetcode id=290 lang=python3
#
# [290] Word Pattern
#

# @lc code=start

# * Hash Table Solution | O(n) Time | O(n) Space
# * n -> The length of s or pattern string


class Solution:
    def wordPattern(self, pattern: str, s: str) -> bool:
        words = s.split()
        if len(pattern) != len(words) or len(set(pattern)) != len(set(words)):
            return False

        char_word_map = {}
        for char, word in zip(pattern, words):
            if char in char_word_map and char_word_map[char] != word:
                return False

            char_word_map[char] = word

        return True


# @lc code=end
